source/common/conn_pool/conn_pool_base.cc

```diff
@@ -891,12 +891,16 @@ void ActiveClient::onConnectionDurationTimeout() {
 }
 
 void ActiveClient::drain() {
-  if (currentUnusedCapacity() <= 0) {
-    return;
-  }
-  parent_.decrConnectingAndConnectedStreamCapacity(currentUnusedCapacity(), *this);
+  const int64_t unused = currentUnusedCapacity();
 
+  // Remove draining client's capacity from the pool. In the case that it currently has
```

Comment:
This is the minimum change to fix this issue. However, I think it might be clearer if, when a connection starts draining, we remove it's currentUnusedCapacity from the pool/cluster state, and never incr/decr for a pool in state draining. @botengyao wdyt?

Making that change would mean that negative draining capacity would no longer be counted in pool capacity, which I think is better than current behavior, but is a slight behavior change.

Reply:
I think the behavior you described makes more sense and I agree that there is something we need to do better and fix around the draining process and we need to avoid the add-back when it is draining.

Regarding this logic, how does the `remaining_streams_ = 0` prevent the add back? I am reading this code:

https://github.com/envoyproxy/envoy/blob/18f98970d6b002468da708c73a505892de48824f/source/common/conn_pool/conn_pool_base.cc#L276-L288

and it seems it is still checking both `negative_capacicy` and `limited_by_concurrency`. Could we change something like below that probably you already mentioned as a clearer way when a request is finished.

```
if (negative && client.state() != ActiveClient::State::Draining) {
```

Reply:
If `remaining_capacity_` isn't set to zero, then after `currentUnusedCapacity()` stops being negative, the connection is considered `limited_by_concurrency`, so all of it's available capacity is added back into the pool. But this is incorrect; it's limited by state draining, not by concurrency limits.

Reply:
Yes, the first version in this PR correctly fixed so that any non-negative capacity on the draining connection is not added into pool capacity. The current/second version also removes negative capacity from the pool for draining connections. This will mean fewer connections are computed as required.

As you noted in your examples above, the computed capacity will be larger (right hand side of inequality) because it  may not count some negative capacity. I think this is more accurate, but will result in fewer connections in this situation.

But if you prefer the previous version of this PR, I'm fine with that. It's a smaller change.

Reply:
Then I prefer the original version, and could you add more comments to explain the behavior around the `remain_streams_ = 0` changes, it is not that explicit. Thank you!